### PR TITLE
fix a bug when deleting rocksdb db_  pointer, there is a segmentfault

### PR DESCRIPTION
**Pull request by LLiuJJ**
使用 unique_ptr 会导致 delete rocksdb db 指针的时候 core 掉，下面是发生 core 具体的调用信息。

liuj@fastdb:~/corefile$ gdb qedis_server core-qedis_server-34045-1608299232
GNU gdb (Ubuntu 9.1-0ubuntu1) 9.1
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from qedis_server...

warning: exec file is newer than core file.
[New LWP 34045]
[New LWP 34051]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `./qedis_server qedis.conf'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  rocksdb::ThreadPoolImpl::Impl::UnSchedule (this=0x55848b8046b0, arg=0x55848b806990)
    at util/threadpool_imp.cc:401
401	      if (arg == (*it).tag) {
[Current thread is 1 (Thread 0x7f1196f00ac0 (LWP 34045))]
(gdb) where
#0  rocksdb::ThreadPoolImpl::Impl::UnSchedule (this=0x55848b8046b0, arg=0x55848b806990)
    at util/threadpool_imp.cc:401
#1  0x00007f11977c3aac in rocksdb::DBImpl::CloseHelper (this=this@entry=0x55848b806990)
    at db/db_impl/db_impl.cc:452
#2  0x00007f11977cc5fd in rocksdb::DBImpl::~DBImpl (this=0x55848b806990, __in_chrg=<optimized out>)
    at db/db_impl/db_impl.cc:608
#3  0x00007f11977cc711 in rocksdb::DBImpl::~DBImpl (this=0x55848b806990, __in_chrg=<optimized out>)
    at db/db_impl/db_impl.cc:610
#4  0x00007f1197c38a29 in qedis::QRocksdb::~QRocksdb (this=0x55848b803680, __in_chrg=<optimized out>)
    at /home/liuj/Documents/Qedis/QedisCore/QRocksdb.cc:20
#5  0x00007f1197c4cece in std::default_delete<qedis::QDumpInterface>::operator() (this=<optimized out>, 
    __ptr=<optimized out>) at /usr/include/c++/7/bits/unique_ptr.h:78
#6  std::unique_ptr<qedis::QDumpInterface, std::default_delete<qedis::QDumpInterface> >::~unique_ptr (
    this=<optimized out>, __in_chrg=<optimized out>) at /usr/include/c++/7/bits/unique_ptr.h:263
#7  std::_Destroy<std::unique_ptr<qedis::QDumpInterface, std::default_delete<qedis::QDumpInterface> > > (
    __pointer=<optimized out>) at /usr/include/c++/7/bits/stl_construct.h:98
#8  std::_Destroy_aux<false>::__destroy<std::unique_ptr<qedis::QDumpInterface, std::default_delete<qedis::QDumpInterface> >*> (__last=<optimized out>, __first=0x55848b821050) at /usr/include/c++/7/bits/stl_construct.h:108
#9  std::_Destroy<std::unique_ptr<qedis::QDumpInterface, std::default_delete<qedis::QDumpInterface> >*> (
    __last=<optimized out>, __first=<optimized out>) at /usr/include/c++/7/bits/stl_construct.h:137
#10 std::_Destroy<std::unique_ptr<qedis::QDumpInterface, std::default_delete<qedis::QDumpInterface> >*, std::unique_ptr<qedis::QDumpInterface, std::default_delete<qedis::QDumpInterface> > > (__last=0x55848b821058, 
    __first=<optimized out>) at /usr/include/c++/7/bits/stl_construct.h:206
#11 std::vector<std::unique_ptr<qedis::QDumpInterface, std::default_delete<qedis::QDumpInterface> >, std::allocator<std::unique_ptr<qedis::QDumpInterface, std::default_delete<qedis::QDumpInterface> > > >::~vector (
    this=0x7f1197c6f928 <qedis::QStore::Instance()::store+72>, __in_chrg=<optimized out>)
    at /usr/include/c++/7/bits/stl_vector.h:434
#12 qedis::QStore::~QStore (this=0x7f1197c6f8e0 <qedis::QStore::Instance()::store>, __in_chrg=<optimized out>)
    at /home/liuj/Documents/Qedis/QedisCore/QStore.h:82
#13 0x00007f1197241a27 in __run_exit_handlers (status=0, listp=0x7f11973e3718 <__exit_funcs>, 
    run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#14 0x00007f1197241be0 in __GI_exit (status=<optimized out>) at exit.c:139
#15 0x00007f119721f0ba in __libc_start_main (main=0x5584899ad6a0 <main(int, char**)>, argc=2, 
    argv=0x7ffd08266778, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, 
--Type <RET> for more, q to quit, c to continue without paging--
    08266768) at ../csu/libc-start.c:342
#16 0x00005584899ad8be in _start ()

修改智能指针为普通指针来修复问题，在程序停止前手动回收内存。

### CODE REVIEW

**Comment by loveyacper** [comment]
确定是因为智能指针带来的问题吗？感觉像是没有优雅关闭。
能否在关闭rocksdb前，执行一些类似closeDB这样的操作再析构对象。

File: QedisCore/QStore.cc [hunk 1]
```diff
@@ -670,6 +670,15 @@ void QStore::ResetDb()
     dbno_ = 0;
 }
 
+void QStore::FreeAllBackends()
+{
+    for (std::vector<QDumpInterface* >::iterator it = backends_.begin(); it != backends_.end(); ++it)
```
Comment (loveyacper) [inline QedisCore/QStore.cc:675]: 可以保持unique_ptr不变。这里只要backends_.clear();就可以了，提前触发leveldb/rocksdb的析构

**Comment by loveyacper** [comment]
可以考虑在FreeAllBackends函数中，直接执行backends_.clear()去触发db析构。不必要取消unique-ptr。

**Comment by LLiuJJ** [comment]
发件人: Bert Young ***@***.***>
答复: loveyacper/Qedis ***@***.***>
日期: 2021年3月26日 星期五 下午3:53
收件人: loveyacper/Qedis ***@***.***>
抄送: Colin ***@***.***>, Author ***@***.***>
主题: Re: [loveyacper/Qedis] fix a bug when deleting rocksdb db_ pointer, there is a segmentfault (#11)

unique_ptr

辛苦指点，我改一下测下

**Pull request closed without merging**